### PR TITLE
P20-424 Adding start_libraries to i18n sync

### PR DESCRIPTION
**Pull request by mgc1194**
In order to support translatability in Sample Apps, apps where student code is hidden and all students need to do is interact with the app, (**just** _sample Apps_, **NOT** _starter Apps_), @dancodedotorg has found a workaround by using student.
Curriculum writers would create a library containing a JavaScript dictionary containing key-string pairs (that can be parsed as a json object) and a translate function that returns the string for a given key. See attached image.
image

This workaround is similar to our current translation system for JS/React

Curriculum writers can export these libraries, and import them in the level apps.
translation-poc

When a level has imported libraries, those get stored in the `start_libraries` property of the level (`level.start_libraries`).

start_libraries get distributed in the sync out as the rest of the curriculum content:

Screenshot 2023-10-03 at 09 16 44

This PR does:
- Expose start_libraries as a serialized attribute for levels.
- For levels with start libraries, check if any of the libraries are I18n-related.
- Collect the strings from `TRANSLATIONTEXT` add them to level strings and get them into the I18n sync pipeline.

This PR does not:
- Updates `blockly.rb` model to translate `start_libraries`.

## Links

- jira ticket: [P20-424](https://codedotorg.atlassian.net/browse/P20-424)

## Testing story
Manually ran the sync and check start libraries were uploaded to Crowdin and translated content was distributed to Dashboard:
Screenshot 2023-09-29 at 23 56 26

## Follow-up work

Edit `blockly.rb` to substitute English strings by translated ones.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: bin/i18n/resources/dashboard/course_content/sync_in.rb [hunk 1]
```diff
@@ -297,6 +304,7 @@ def prepare_level_content
                 # Don't localize encrypted levels; the whole point of encryption is to
                 # keep the contents of certain levels secret.
                 next if level.encrypted?
+                next unless level.start_libraries
```
Comment (mgc1194) [inline bin/i18n/resources/dashboard/course_content/sync_in.rb]: Don't forget to remove this!

File: bin/i18n/resources/dashboard/course_content/sync_in.rb [hunk 2]
```diff
@@ -155,14 +155,21 @@ def get_i18n_strings(level)
...
+                  library_name = library["name"]
+                  library_source = library["source"]
+                  translation_text = library_source.match(/var TRANSLATIONTEXT = (\{[^}]*});/m).captures
+
+                  i18n_strings['start_libraries'] = Hash.new if i18n_strings['start_libraries'].blank?
+                  i18n_strings['start_libraries'][library_name] = {}
+
+                  JSON.parse(translation_text[0]).each do |key, value|
+                    i18n_strings['start_libraries'][library_name][key] = value
+                  end
```
Comment (artem-vavilov) [inline bin/i18n/resources/dashboard/course_content/sync_in.rb]: Should we check if `library['name']` and `library['source']` are present?
```suggestion
              if level.start_libraries.present?
                level_libraries = JSON.parse(level.start_libraries)
                level_libraries.each do |library|
                  library_name = library['name']
                  next if library_name.blank?
                  next unless /^i18n_/i.match?(library_name)

                  library_source = library['source']
                  next if library_source.blank?

                  translation_json = library_source[/var TRANSLATIONTEXT = (\{[^}]*});/m, 1]
                  next if translation_json.blank?

                  i18n_strings['start_libraries'] ||= {}
                  i18n_strings['start_libraries'][library_name] = {}

                  JSON.parse(translation_json).each do |key, value|
                    i18n_strings['start_libraries'][library_name][key] = value
                  end
```

**Review by artem-vavilov — APPROVED** [review]
Great work!

File: bin/i18n/resources/dashboard/course_content/sync_in.rb (hunk 2 above)
Comment (mgc1194) [reply bin/i18n/resources/dashboard/course_content/sync_in.rb]: Libraries can not be exported if they have no source:
Screenshot 2023-10-02 at 19 15 47

In addition, you can not leave the library name blank:

https://github.com/code-dot-org/code-dot-org/assets/66776217/a69290ab-612b-4481-b030-f2bb5e1bd611

File: bin/i18n/resources/dashboard/course_content/sync_in.rb [hunk 3]
```diff
@@ -155,14 +155,21 @@ def get_i18n_strings(level)
...
-                i18n_strings['start_html'] = Hash.new unless level.start_html.empty?
-
-                # match any element that contains text
-                start_html.xpath('//*[text()[normalize-space()]]').each do |element|
-                  i18n_strings['start_html'][element.text] = element.text
+              # start_libraries
+              unless level.start_libraries.blank?
+                level_libraries = JSON.parse(level.start_libraries)
+                level_libraries.each do |library|
+                  next unless /^i18n_/i.match?(library["name"])
```
Comment (daynew) [inline bin/i18n/resources/dashboard/course_content/sync_in.rb]: Is there a plan to have multiple names? If there are just 1-3 libraries, I would prefer an explicit list. It makes it easier for future developers who grep the codebase looking for all the locations "i18n_starter_library" is used.

File: bin/i18n/resources/dashboard/course_content/sync_in.rb (hunk 2 above)
Comment (daynew) [reply bin/i18n/resources/dashboard/course_content/sync_in.rb]: Although that's true now, it's possible a legacy level might have blank values. You could do a quick query to verify this or you could add the blank checks.

**Review by daynew — APPROVED** [review]
Looks good, just a couple comments

File: bin/i18n/resources/dashboard/course_content/sync_in.rb (hunk 2 above)
Comment (mgc1194) [reply bin/i18n/resources/dashboard/course_content/sync_in.rb]: It is worth checking.

File: bin/i18n/resources/dashboard/course_content/sync_in.rb (hunk 3 above)
Comment (mgc1194) [reply bin/i18n/resources/dashboard/course_content/sync_in.rb]: The plan for these libraries is to be added as needed to support sample apps, with a set of guidelines agreed with curriculum writers.
Libraries will be shared by multiple levels in a unit, but they have limitations in size so we can't have infinite long libraries.

File: bin/i18n/resources/dashboard/course_content/sync_in.rb (hunk 2 above)
Comment (mgc1194) [reply bin/i18n/resources/dashboard/course_content/sync_in.rb]: After querying the levels with libraries and talking with @jmkulwik, Libraries are required to have a name and at least one function.
Old libraries that don't follow this format can be deleted as they would not work.

**Pull request merged**